src/oneelement.jl

```diff
@@ -93,7 +93,56 @@ function *(A::OneElementMatrix, B::AbstractFillVector)
     OneElement(val, A.ind[1], size(A,1))
 end
 
-@inline function __mulonel!(y, A, x, alpha, beta)
+# Special matrix types
+
+function *(A::OneElementMatrix, D::Diagonal)
+    check_matmul_sizes(A, D)
+    nzcol = A.ind[2]
+    val = if nzcol in axes(D,1)
+        A.val * D[nzcol, nzcol]
+    else
+        A.val * zero(eltype(D))
+    end
+    OneElement(val, A.ind, size(A))
+end
+function *(D::Diagonal, A::OneElementMatrix)
+    check_matmul_sizes(D, A)
+    nzrow = A.ind[1]
+    val = if nzrow in axes(D,2)
+        D[nzrow, nzrow] * A.val
+    else
+        zero(eltype(D)) * A.val
+    end
+    OneElement(val, A.ind, size(A))
+end
+
+# Inplace multiplication
+
+function __mulonel!(C, A, B, alpha, beta)
```

Comment:
```suggestion
# We use this for out overloads for _mul! for OneElement because its more efficient
# due to how efficient 2 arg mul is when one or more of the args are OneElement
function __mulonel!(C, A, B, alpha, beta)
```